### DISCUSSION TITLE
Need to know which are cultures are supported for TextBlock.IsHyphenationEnabled

### DISCUSSION THREAD

**Discussion by ankushmadankar** [dotnet/wpf#9731] [General]
I'm currently working on a WPF application and using the TextBlock.IsHyphenationEnabled property to handle text hyphenation. I understand that hyphenation behavior can be culture-specific, and I'm trying to ensure proper functionality across different languages.

Could anyone provide a list of cultures that are officially supported for hyphenation when using the `TextBlock.IsHyphenationEnabled` property? Specifically, I'm interested in knowing if languages like Swedish (sv-SE), German (de-DE), and others are fully supported.

I see some code related to it but not sure how culture support is done:

Class: `NaturalLanguageHyphenator`

```
public override bool IsCultureSupported(CultureInfo culture)
{
    // Accept all cultures for the time being. Ideally, NL6 should provide a way for the client to test supported culture.
    return true;
}
```
Additionally, if any specific settings or configurations are required on the client machine, such as installing language packs in Windows.

Code and debugging values for TextBlock and Culture info
```
<TextBlock IsHyphenationEnabled="True" 
    Text="{Binding Name}" Language="sv-SE"
    TextWrapping="Wrap"/>
```
	
```
  System.Globalization.CultureInfo.CurrentCulture - en-IN
  System.Globalization.CultureInfo.CurrentUICulture - sv
```

Machine with language sv-SE installed is not able to hyphenate string.

Any documentation or experiences you can share would be greatly appreciated!

Thanks in advance for your help.

**Comment by miloush** [comment]
The hyphenation is performed by the OS. The supported languages can differ based OS version and indeed installed language features.

Depending on the application, you might want the user to be able to enable/disable hyphenation, but I would expect if you enable it and there is no support for given language, the text will not be hyphenated.

**Reply by ankushmadankar** [reply]
Hello miloush, Thank you for your response. I have this OS and Lang setting and hyphenation is not working for sv-SE. Would like to suggest something which I can try? I have updated code and culture info values

OS Name: Microsoft Windows 11 Pro
OS Version: 10.0.22631 N/A Build 22631
Default system UI language : en-US
System locale : en-US
Installed language(s): en-GB
  Type : Partially localized language, MUI type.
  Fallback Languages en-US
Installed language(s): en-US
  Type : Fully localized language.
Installed language(s): sv-SE
  Type : Partially localized language, MUI type.
  Fallback Languages en-US

**Reply by miloush** [reply]
Are yo able to provide a repro app for me to try?

**Reply by ankushmadankar** [reply]
Hello miloush, I have created sample wpf & console app for hyphenation. Please check readme for brief intro. https://github.com/ankushmadankar/hyphenation

**Reply by miloush** [reply]
To answer your question, the .NET Framework ships with English, German, French and Spanish. A list of others once supported by OS is available here: https://learn.microsoft.com/en-us/archive/blogs/wpf/wpf-in-net-4-6-1. These at some point got moved into optional, on-demand components, but I confirm that on Windows 11, installing language pack or input language does not install the required data. I am trying to figure out whether there is another way to install it.

**Reply by ankushmadankar** [reply]
I am using dotnet 6 and not .NET Framework, will this behavior same for dotnet 6? Is it possible to do hyphenation for language German without adding additional components (dotnet 6 or dotnet 8)?

**Reply by miloush** [reply]
No and after some more investigation it looks like the .NET Framework information is also outdated and it now ships with English only, although when I target .NET Framework, it loads data from OS anyway, so it doesn't matter.

What you need is NlsData001d.dll in the system directory, where 001d is the lang id for Swedish etc.

**Reply by ankushmadankar** [reply]
I have NlsData001d.dll in system directory (C:\Windows\System32), still hyphenation not working for sv-SE! Where to find lang id e.g. 001d, is it LCID? Surprisingly Hyphenation at all not working for machines which are running on Swedish culture even after setting en-US application language, I checked with console app which I have created where it is not able find UnsafeNativeMethods.NlCreateHyphenator. Is anything additional data I need setup or hyphenation is currently broken?

**Reply by miloush** [reply]
001d is LANG ID which is part of LCID. See https://learn.microsoft.com/en-us/windows/win32/intl/language-identifiers and especially https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-lcid/70feba9f-294e-491e-b6eb-56532684c37f.

It looks like the NLS data does get installed when adding an input language, but not all of them support hyphenation. I think I have run out of time I can spend on this. Someone from OS needs to tell us which languages are supposed to be supported (which is the original question) and then we can see if something on WPF side is broken, which is unlikely, given that it just calls OS for the heavy work. Sorry I couldn't help more!
